### PR TITLE
Add ctx.send* methods and optional useNewReplies from telegraf/future

### PR DESCRIPTION
**Pull request by MKRhere**
It has been a long-confusing feature of Telegraf that `ctx.reply` doesn't actually reply, but sends a message in the same chat. We now plan to fix it. This PR introduces `ctx.sendMessage`, `ctx.sendPhoto`, etc., that reflect the upstream API, but implicitly send a message in the same chat, more consistently with the other Context methods. This is the current behaviour of the `ctx.reply` methods.

This PR also introduces the new reply mode, which is optionally enabled using the `useNewReplies` middleware. Demonstration:

```TS
import { Telegraf } from "telegraf";
import { useNewReplies } from "telegraf/future";

const bot = new Telegraf(token);

// current behaviour:

bot.on("message", ctx => {
  // sends message in the same group, but doesn't reply
  ctx.reply("Hello");

  // sends message in the same group, but doesn't reply
  ctx.sendMessage("Hello");
});

bot.use(useNewReplies());

// new behaviour:

bot.on("message", ctx => {
  // actually replies to the message that was received
  ctx.reply("Hello");

  // sends message in the same group, but doesn't reply
  ctx.sendMessage("Hello");
});
```

This causes `ctx.reply`, `ctx.replyWithPhoto`, and the like to actually reply to the message in current context. If a message is not available, it will simply send a regular message.

A future minor release of telegraf will deprecate the current `ctx.reply*` methods in favour of the new ones, causing their usage to log warnings, prompting users to enable `useNewReplies`. v5 will completely replace the behaviour of `ctx.reply` with the new reply methods, ~~and an optional middleware, `useOldReplies`, to replace them with the old methods may be added.~~

### CODE REVIEW

File: src/future.ts
Comment (wojpawlik) [inline src/future.ts]: Avoid mutating users' arguments:
```suggestion
  const reply_to_message_id = ctx.message?.message_id
  return { reply_to_message_id, ... extra }
```

File: src/context.ts
Comment (wojpawlik) [inline src/context.ts]:
```suggestion
    this.assert(this.chat, 'sendPhoto')
```

File: src/context.ts
Comment (wojpawlik) [inline src/context.ts]:
```suggestion
    this.assert(this.chat, 'sendMediaGroup')
```

File: src/context.ts
Comment (wojpawlik) [inline src/context.ts]:
```suggestion
    this.assert(this.chat, 'sendSticker')
```

File: src/context.ts
Comment (wojpawlik) [inline src/context.ts]:
```suggestion
    this.assert(this.chat, 'sendAnimation')
```

File: src/util.ts
Comment (wojpawlik) [inline src/util.ts]: Why not just make `Context::assert` `@internal public` and consistently use it everywhere?

File: src/util.ts
Comment (wojpawlik) [inline src/util.ts:25]:
```suggestion
  process.emitWarning(`${method} is deprecated, use ${use} instead.`, 'DeprecationWarning')
```

**Review by wojpawlik — CHANGES_REQUESTED** [review]
The history is a little messy already, I suggest merge-squashing or rebasing before merging.

> an optional middleware, `useOldReplies`, to replace them with the old methods may be added.

Just throw away the old behavior, that's the point of a major release.

File: src/util.ts
Comment (MKRhere) [reply src/util.ts:25]: `process` is node-specific; we don't want telegraf to be more locked into node in the future

File: src/util.ts
Comment (MKRhere) [reply src/util.ts]: What's `@internal`?

Comment (wojpawlik) [reply src/util.ts]: https://www.typescriptlang.org/tsconfig#stripInternal

**Comment by KnorpelSenf** [comment]
I will review this in a few days, my time is currently a bit limited. Please don't wait for me if prefer to merge before I get to this.

**Comment by wojpawlik** [comment]
I see no reason to rush this

**Comment by MKRhere** [comment]
@wojpawlik 

> I suggest merge-squashing or rebasing before merging

I have rebased, but prefer not to squash history. Rather than treat git as a changelog, I prefer to treat it as a devlog.

> Just throw away the old behavior

Okay, I've updated the PR description to reflect this

@KnorpelSenf 

> my time is currently a bit limited

No pressure; I might merge after some more testing to make sure nothing is breaking. If you get time meanwhile, appreciate a review.

**Review by wojpawlik — APPROVED** [review]

**Comment by MKRhere** [comment]
This PR is ready to merge into feat-api-6.2, and I'll bundle them into one release at 4.9.0. However, I'm not merging yet so as to not clog reviews on #1671. When #1671 is approved and ready to be merged, I'll merge this first.

**Comment by wojpawlik** [comment]
Why not merge this into v4 right now?

**Pull request merged**
